File: backend/app/schemas/delivery.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import (
    DeliveryFailureReason,
    DeliveryProviderType,
    DeliveryStatus,
    ProofOfDeliveryType,
)


class DeliveryProviderRead(BaseModel):
    id: UUID
    code: str
    name: str
    provider_type: DeliveryProviderType
    active: bool
    api_enabled: bool

    model_config = {"from_attributes": True}


class DeliveryAgentCreate(BaseModel):
    provider_id: UUID
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    phone: str = Field(min_length=1, max_length=30)
    email: str | None = None


class DeliveryAgentUpdate(BaseModel):
    active: bool


class DeliveryAgentRead(BaseModel):
    id: UUID
    provider_id: UUID
    first_name: str
    last_name: str
    phone: str
    email: str | None
    active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class DeliveryAttemptRead(BaseModel):
    id: UUID
    attempt_number: int
    attempted_at: datetime
    courier_id: UUID | None
    reason: DeliveryFailureReason
    notes: str | None

    model_config = {"from_attributes": True}


class ProofOfDeliveryRead(BaseModel):
    delivered_at: datetime
    delivered_by: str
    recipient_name_if_provided: str | None
    delivery_method: str | None
    proof_type: ProofOfDeliveryType
    notes: str | None

    model_config = {"from_attributes": True}


class DeliveryOrderSummary(BaseModel):
    id: UUID
    tracking_number: str
    letter_reference: str | None
    recipient_name: str
    status: DeliveryStatus
    provider_name: str
    courier_name: str | None
    created_at: datetime
    updated_at: datetime


class DeliveryOrderRead(BaseModel):
    id: UUID
    letter_id: UUID
    letter_reference: str | None
    tracking_number: str
    status: DeliveryStatus
    provider: DeliveryProviderRead
    courier: DeliveryAgentRead | None
    attempt_count: int
    created_at: datetime
    assigned_at: datetime | None
    picked_up_at: datetime | None
    in_transit_at: datetime | None
    out_for_delivery_at: datetime | None
    delivered_at: datetime | None
    failed_at: datetime | None
    returned_at: datetime | None
    updated_at: datetime
    attempts: list[DeliveryAttemptRead]
    proof: ProofOfDeliveryRead | None


class AssignCourierRequest(BaseModel):
    agent_id: UUID


class DeliveryFailureRequest(BaseModel):
    reason: DeliveryFailureReason
    notes: str | None = Field(default=None, max_length=2000)


class DeliveryConfirmRequest(BaseModel):
    notes: str | None = Field(default=None, max_length=2000)


class DeliveryPublicEvent(BaseModel):
    label: str
    at: datetime


class DeliveryPublicView(BaseModel):
    """Safe subset shown on the public tracking page and the recipient's
    secure-link view -- no courier personal info, no admin notes, no
    internal IDs (spec section 27)."""

    tracking_number: str
    status: DeliveryStatus
    events: list[DeliveryPublicEvent]

File: frontend/src/types/delivery.ts
export type DeliveryStatus =
  | "CREATED"
  | "READY_FOR_DISPATCH"
  | "ASSIGNED"
  | "PICKED_UP"
  | "IN_TRANSIT"
  | "OUT_FOR_DELIVERY"
  | "DEPOSITED"
  | "DELIVERED"
  | "DELIVERY_FAILED"
  | "RETURNED_TO_SENDER"
  | "CANCELLED"

export type DeliveryFailureReason =
  | "RECIPIENT_UNAVAILABLE"
  | "INCORRECT_ADDRESS"
  | "RECIPIENT_REFUSED"
  | "ADDRESS_INACCESSIBLE"
  | "OTHER"

export interface DeliveryProviderRead {
  id: string
  code: string
  name: string
  provider_type: "INTERNAL" | "EXTERNAL_CARRIER" | "POSTAL_SERVICE"
  active: boolean
  api_enabled: boolean
}

export interface DeliveryAgentRead {
  id: string
  provider_id: string
  first_name: string
  last_name: string
  phone: string
  email: string | null
  active: boolean
  created_at: string
}

export interface DeliveryAttemptRead {
  id: string
  attempt_number: number
  attempted_at: string
  courier_id: string | null
  reason: DeliveryFailureReason
  notes: string | null
}

export interface ProofOfDeliveryRead {
  delivered_at: string
  delivered_by: string
  recipient_name_if_provided: string | null
  delivery_method: string | null
  proof_type: string
  notes: string | null
}

export interface DeliveryOrderSummary {
  id: string
  tracking_number: string
  letter_reference: string | null
  recipient_name: string
  status: DeliveryStatus
  provider_name: string
  courier_name: string | null
  created_at: string
  updated_at: string
}

export interface DeliveryOrderRead {
  id: string
  letter_id: string
  letter_reference: string | null
  tracking_number: string
  status: DeliveryStatus
  provider: DeliveryProviderRead
  courier: DeliveryAgentRead | null
  attempt_count: number
  created_at: string
  assigned_at: string | null
  picked_up_at: string | null
  in_transit_at: string | null
  out_for_delivery_at: string | null
  deposited_at: string | null
  delivered_at: string | null
  failed_at: string | null
  returned_at: string | null
  updated_at: string
  attempts: DeliveryAttemptRead[]
  proof: ProofOfDeliveryRead | null
}

export interface DeliveryPublicEvent {
  label: string
  at: string
}

export interface DeliveryPublicView {
  tracking_number: string
  status: DeliveryStatus
  events: DeliveryPublicEvent[]
}

export interface DeliveryConfirmationView {
  reference: string
  tracking_number: string
  sender_first_name: string
  sender_last_name: string
  confirmed: boolean
}
